src/plugins/competition/proposal.ts

```diff
@@ -0,0 +1,336 @@
+import { IProposalState, Proposal } from "../proposal"
+import { IEntityRef } from "../../entity"
+import { Arc, IApolloQueryOptions } from "../../arc"
+import { secondSinceEpochToDate } from "../../utils"
+import { Address } from "../../types"
+import { Observable, from } from "rxjs"
+import gql from "graphql-tag"
+import { CompetitionVote } from "./vote"
+import { ITransactionReceipt, getEventArgs, ITransaction, toIOperationObservable, Operation } from "../../operation"
+import { CompetitionSuggestion } from "./suggestion"
+import { DAO } from "../../dao"
+import { first, concatMap } from "rxjs/operators"
+import { Competition } from "./plugin"
+import { Plugin } from "../plugin"
+import { IContributionRewardExtState } from "../contributionRewardExt/plugin"
+
+export interface ICompetitionProposalState extends IProposalState { 
+  id: string
+  admin: Address
+  contract: Address
+  endTime: Date
+  numberOfWinners: number
+  rewardSplit: number[]
+  startTime: Date
+  votingStartTime: Date
+  suggestionsEndTime: Date
+  numberOfVotesPerVoter: number
+  plugin: IEntityRef<Competition>
+  snapshotBlock: number
+  createdAt: Date
+  totalVotes: number
+  totalSuggestions: number
+  numberOfWinningSuggestions: number
+}
+
+export class CompetitionProposal extends Proposal {
+
+  public static fragment = {
+    name: 'CompetitionProposalFields',
+    fragment: gql`
+      fragment CompetitionProposalFields on Proposal {
+        competition {
+          id
+          admin
+          endTime
+          contract
+          suggestionsEndTime
+          createdAt
+          numberOfWinningSuggestions
+          numberOfVotesPerVoters
+          numberOfWinners
+          rewardSplit
+          snapshotBlock
+          startTime
+          totalSuggestions
+          totalVotes
+          votingStartTime
+  
+        }
+      }
+    `
+  }
+
+  coreState: ICompetitionProposalState| undefined
```

Comment:
Isn't this duplicating what the base class is doing?

Reply:
We should probably templatize the Proposal class, so we can do `class CompetitionProposal extends Proposal<ICompetitionProposalState>`.

Reply:
So happy to see comprehensive use of types!  :-)